vscode/src/testExplorer.ts

```diff
@@ -0,0 +1,267 @@
+ * if `shouldDeleteOldTests` is `true`, then clear out previously discovered tests. If `false`, add new tests but don't dissolve old ones.
+ *
+ */
+function mkRefreshHandler(
+  ctrl: vscode.TestController,
+  context: vscode.ExtensionContext,
+  shouldDeleteOldTests: boolean = true,
+) {
+  return async () => {
+    if (shouldDeleteOldTests) {
+      for (const [id] of ctrl.items) {
+        ctrl.items.delete(id);
+      }
+    }
+    const programConfig = await getProgramConfig();
+    if (!programConfig) {
+      return;
+    }
+    const worker = localGetCompilerWorker(context);
+
+    const testCallables = await worker.collectTestCallables(programConfig);
+
+    // break down the test callable into its parts, so we can construct
+    // the namespace hierarchy in the test explorer
+    for (const testCallable of testCallables) {
+      const parts = testCallable.split(".");
+
+      // for an individual test case, e.g. foo.bar.baz, create a hierarchy of items
+      let rover = ctrl.items;
+      for (let i = 0; i < parts.length; i++) {
+        const part = parts[i];
+        const id = i === parts.length - 1 ? testCallable : part;
+        if (!rover.get(part)) {
+          rover.add(ctrl.createTestItem(id, part));
+        }
+        rover = rover.get(id)!.children;
+      }
+    }
+  };
+}
+
+export async function initTestExplorer(context: vscode.ExtensionContext) {
+  const ctrl: vscode.TestController = vscode.tests.createTestController(
+    "qsharpTestController",
+    "Q# Tests",
+  );
+  context.subscriptions.push(ctrl);
+  // construct the handler that runs when the user presses the refresh button in the test explorer
+  const refreshHandler = mkRefreshHandler(ctrl, context);
+  // initially populate tests
+  await refreshHandler();
+
+  ctrl.refreshHandler = refreshHandler;
+
+  const runHandler = (request: vscode.TestRunRequest) => {
+    if (!request.continuous) {
+      return startTestRun(request);
+    }
+  };
+
+  // runs an individual test run
+  // or test group (a test run where there are child tests)
+  const startTestRun = async (request: vscode.TestRunRequest) => {
+    // use the compiler worker to run the test in the interpreter
+
+    log.info("Starting test run, request was", JSON.stringify(request));
+    const worker = localGetCompilerWorker(context);
+
+    const program = await getProgramConfig();
+    if (!program) {
+      return;
+    }
+
+    for (const testCase of request.include || []) {
+      await runTestCase(ctrl, testCase, request, worker, program);
+    }
+  };
+
+  ctrl.createRunProfile(
+    "Run Tests",
+    vscode.TestRunProfileKind.Run,
+    runHandler,
+    true,
+    undefined,
+    false,
+  );
+
+  ctrl.resolveHandler = async (item) => {
+    if (!item) {
+      context.subscriptions.push(...startWatchingWorkspace(ctrl, context));
+      return;
+    }
+  };
+
+  function updateNodeForDocument(e: vscode.TextDocument) {
+    if (e.uri.scheme !== "file") {
```

Comment:
I believe this means Test Explorer won't work for the playground (which has its own virtual file system), perhaps GitHub (which has it's own scheme) and maybe others. Any reason to restrict this to 'file' URIs?

Reply:
I think the virtual file system and the URI scheme are separate; things in a vfs are still scheme `file`. It works as expected in the browser playground for me:
<img width="1198" alt="image" src="https://github.com/user-attachments/assets/eed45fe2-f7de-4bb9-8cd9-1f40c5aed324" />

Reply:
By restricting to `file` scheme it makes sure tests are detected from things like diffs or files open from PRs (the fixes in #1754 and #1867 respectively).

Reply:
Actually, looking at those old fixes, I wonder if it would be best to use `isQsharpDocument` from common for this, since we already rely on that logic for reporting errors and other things.

Reply:
(in fact in the url in your screen shot you can see the first param for the folder to open starts with `qsharp-vfs://` (url encoded). I'll figure it out and report back. Might be worth a wiki page on how alternate schemes like playground and github are handled for future reference)

Reply:
> Actually, looking at those old fixes, I wonder if it would be best to use isQsharpDocument from common for this

Good call. And there we explicitly excluded schemes we know to be problematic, which allows us to be flexible for other schemes we _should_ be able to handle.

Reply:
I can switch to that. I was following the patterns used in the sample from the VS Code docs. Good call on using `!isQsharpDocument` -- updated.